File: src/fossilize.js
// ============================================================
// fossilize.js — Freeze a Prolog engine's clause database
// mineralize.js — Selectively lock specific predicates
//
// Two modes of hardening:
//
//   mineralize(engine, "react", 0)
//     → react/0 is immutable.  Everything else stays dynamic.
//     → Additive, one-way.  Call multiple times for multiple preds.
//
//   fossilize(engine)
//     → ALL clauses frozen.  Only ephemeral survives.
//     → Terminal.  Nothing changes after this.
//
// Portable: ES5, no dependencies.
// ============================================================

// ── Helpers ──────────────────────────────────────────────────

function _predKey(term) {
  if (!term) return null;
  if (term.type === "compound") return term.functor + "/" + term.args.length;
  if (term.type === "atom") return term.name + "/0";
  return null;
}

// ── mineralize ──────────────────────────────────────────────

function mineralize(engine, functor, arity) {
  // Initialize mineralized set if needed
  if (!engine.mineralized) {
    engine.mineralized = {};
    _installMineralizeGuards(engine);
  }
  engine.mineralized[functor + "/" + arity] = true;
}

function _installMineralizeGuards(engine) {
  // Guard addClause
  var origAdd = engine.addClause.bind(engine);
  engine.addClause = function(head, body) {
    var key = _predKey(head);
    if (key && engine.mineralized[key]) return;
    origAdd(head, body);
  };

  // Guard retractFirst
  var origRetract = engine.retractFirst.bind(engine);
  engine.retractFirst = function(head) {
    var key = _predKey(head);
    if (key && engine.mineralized[key]) return false;
    return origRetract(head);
  };

  // Guard assert/1
  var origAssert = engine.builtins["assert/1"];
  engine.builtins["assert/1"] = function(goal, rest, subst, counter, depth, onSolution) {
    var term = engine.deepWalk(goal.args[0], subst);
    var key = _predKey(term);
    if (key && engine.mineralized[key]) return; // fail
    origAssert(goal, rest, subst, counter, depth, onSolution);
  };

  // Guard assertz/1 (alias)
  var origAssertz = engine.builtins["assertz/1"];
  if (origAssertz) {
    engine.builtins["assertz/1"] = function(goal, rest, subst, counter, depth, onSolution) {
      var term = engine.deepWalk(goal.args[0], subst);
      var key = _predKey(term);
      if (key && engine.mineralized[key]) return;
      origAssertz(goal, rest, subst, counter, depth, onSolution);
    };
  }

  // Guard retract/1
  var origRetractBuiltin = engine.builtins["retract/1"];
  engine.builtins["retract/1"] = function(goal, rest, subst, counter, depth, onSolution) {
    var term = engine.deepWalk(goal.args[0], subst);
    var key = _predKey(term);
    if (key && engine.mineralized[key]) return;
    origRetractBuiltin(goal, rest, subst, counter, depth, onSolution);
  };

  // Guard retractall/1
  var origRetractAll = engine.builtins["retractall/1"];
  engine.builtins["retractall/1"] = function(goal, rest, subst, counter, depth, onSolution) {
    var term = engine.deepWalk(goal.args[0], subst);
    var key = _predKey(term);
    if (key && engine.mineralized[key]) return;
    origRetractAll(goal, rest, subst, counter, depth, onSolution);
  };

  // Guard ephemeral/1 (if registered)
  if (engine.builtins["ephemeral/1"]) {
    var origEph = engine.builtins["ephemeral/1"];
    engine.builtins["ephemeral/1"] = function(goal, rest, subst, counter, depth, onSolution) {
      var term = engine.deepWalk(goal.args[0], subst);
      var key = _predKey(term);
      if (key && engine.mineralized[key]) return;
      origEph(goal, rest, subst, counter, depth, onSolution);
    };
  }

  // mineralize/1 builtin — callable from Prolog
  engine.builtins["mineralize/1"] = function(goal, rest, subst, counter, depth, onSolution) {
    var term = engine.deepWalk(goal.args[0], subst);
    if (term.type === "compound" && term.functor === "/" && term.args.length === 2) {
      var f = term.args[0];
      var a = term.args[1];
      if (f.type === "atom" && a.type === "num") {
        engine.mineralized[f.name + "/" + a.value] = true;
      }
    }
    engine.solve(rest, subst, counter, depth + 1, onSolution);
  };
}

// ── fossilize ───────────────────────────────────────────────

function fossilize(engine) {
  var boundary = engine.clauses.length;

  // ── Disable permanent mutation builtins (goal fails) ─────
  function _fail() {}  // no solutions — goal fails in Prolog

  engine.builtins["assert/1"] = _fail;
  engine.builtins["assertz/1"] = _fail;
  engine.builtins["retract/1"] = _fail;
  engine.builtins["retractall/1"] = _fail;

  // ── Disable programmatic additions ───────────────────────
  engine.addClause = function() {};

  // ── retractFirst: ephemeral zone only (>= boundary) ──────
  engine.retractFirst = function(head) {
    for (var i = boundary; i < engine.clauses.length; i++) {
      var ch = engine.clauses[i].head;
      var cb = engine.clauses[i].body;
      if (engine.unify(head, ch, new Map()) !== null) {
        engine.clauses.splice(i, 1);
        if (cb.length === 0) {
          for (var j = 0; j < engine.onRetract.length; j++)
            engine.onRetract[j](ch);
        }
        return true;
      }
    }
    return false;
  };

  return boundary;
}

// ── Export (dual ESM/CJS) ───────────────────────────────────

if (typeof exports !== "undefined") {
  exports.fossilize = fossilize;
  exports.mineralize = mineralize;
}
export { fossilize, mineralize };

File: src/fossilize.py
# ============================================================
# fossilize.py — Freeze a Prolog engine's clause database
# mineralize.py — Selectively lock specific predicates
#
# Two modes of hardening:
#
#   mineralize(engine, "react", 0)
#     → react/0 is immutable.  Everything else stays dynamic.
#
#   fossilize(engine)
#     → ALL clauses frozen.  Only ephemeral survives.
# ============================================================

from prolog import unify


def _pred_key(term):
    if term is None:
        return None
    if term[0] == "compound":
        return "%s/%d" % (term[1], len(term[2]))
    if term[0] == "atom":
        return "%s/0" % term[1]
    return None


# ── mineralize ───────────────────────────────────────────────

def mineralize(engine, functor, arity):
    """Lock a specific predicate. One-way, additive."""
    if not hasattr(engine, 'mineralized'):
        engine.mineralized = {}
        _install_mineralize_guards(engine)
    engine.mineralized["%s/%d" % (functor, arity)] = True


def _install_mineralize_guards(engine):
    # Guard add_clause
    orig_add = engine.add_clause

    def _guarded_add(head, body=None):
        key = _pred_key(head)
        if key and key in engine.mineralized:
            return
        orig_add(head, body)

    engine.add_clause = _guarded_add

    # Guard retract_first
    orig_retract = engine.retract_first

    def _guarded_retract(head):
        key = _pred_key(head)
        if key and key in engine.mineralized:
            return False
        return orig_retract(head)

    engine.retract_first = _guarded_retract

    # Guard assert/1
    orig_assert = engine.builtins.get("assert/1")
    if orig_assert:
        def _guarded_assert_builtin(goal, rest, subst, depth, on_sol):
            from prolog import deep_walk
            term = deep_walk(goal[2][0], subst)
            key = _pred_key(term)
            if key and key in engine.mineralized:
                return
            orig_assert(goal, rest, subst, depth, on_sol)
        engine.builtins["assert/1"] = _guarded_assert_builtin

    # Guard assertz/1
    orig_assertz = engine.builtins.get("assertz/1")
    if orig_assertz:
        def _guarded_assertz(goal, rest, subst, depth, on_sol):
            from prolog import deep_walk
            term = deep_walk(goal[2][0], subst)
            key = _pred_key(term)
            if key and key in engine.mineralized:
                return
            orig_assertz(goal, rest, subst, depth, on_sol)
        engine.builtins["assertz/1"] = _guarded_assertz

    # Guard retract/1
    orig_retract_b = engine.builtins.get("retract/1")
    if orig_retract_b:
        def _guarded_retract_builtin(goal, rest, subst, depth, on_sol):
            from prolog import deep_walk
            term = deep_walk(goal[2][0], subst)
            key = _pred_key(term)
            if key and key in engine.mineralized:
                return
            orig_retract_b(goal, rest, subst, depth, on_sol)
        engine.builtins["retract/1"] = _guarded_retract_builtin

    # Guard retractall/1
    orig_retractall = engine.builtins.get("retractall/1")
    if orig_retractall:
        def _guarded_retractall(goal, rest, subst, depth, on_sol):
            from prolog import deep_walk
            term = deep_walk(goal[2][0], subst)
            key = _pred_key(term)
            if key and key in engine.mineralized:
                return
            orig_retractall(goal, rest, subst, depth, on_sol)
        engine.builtins["retractall/1"] = _guarded_retractall

    # Guard ephemeral/1
    orig_eph = engine.builtins.get("ephemeral/1")
    if orig_eph:
        def _guarded_eph(goal, rest, subst, depth, on_sol):
            from prolog import deep_walk
            term = deep_walk(goal[2][0], subst)
            key = _pred_key(term)
            if key and key in engine.mineralized:
                return
            orig_eph(goal, rest, subst, depth, on_sol)
        engine.builtins["ephemeral/1"] = _guarded_eph

    # mineralize/1 builtin — callable from Prolog
    def _mineralize_builtin(goal, rest, subst, depth, on_sol):
        from prolog import deep_walk
        term = deep_walk(goal[2][0], subst)
        if (term[0] == "compound" and term[1] == "/" and len(term[2]) == 2):
            f = term[2][0]
            a = term[2][1]
            if f[0] == "atom" and a[0] == "num":
                engine.mineralized["%s/%d" % (f[1], a[1])] = True
        engine._solve(rest, subst, depth + 1, on_sol)

    engine.builtins["mineralize/1"] = _mineralize_builtin


# ── fossilize ────────────────────────────────────────────────

def fossilize(engine):
    """Freeze the clause database.  Only ephemeral facts allowed after this."""
    boundary = len(engine.clauses)

    def _fail(*args):
        pass

    engine.builtins["assert/1"] = _fail
    engine.builtins["assertz/1"] = _fail
    engine.builtins["retract/1"] = _fail
    engine.builtins["retractall/1"] = _fail

    engine.add_clause = lambda head, body=None: None

    def _fossil_retract_first(head):
        for i in range(boundary, len(engine.clauses)):
            ch, cb = engine.clauses[i]
            s = unify(head, ch, {})
            if s is not None:
                engine.clauses.pop(i)
                if not cb:
                    for fn in engine.on_retract:
                        fn(ch)
                return True
        return False

    engine.retract_first = _fossil_retract_first

    return boundary
